src/USER-BOCS/fix_bocs.cpp

```diff
@@ -630,154 +640,202 @@ void FixBocs::init()
-  fpi = fopen(filename,"r");
-  if (fpi)
-  {
-    while (fgets(line,199,fpi)) { ++n_entries; }
-
-    for (i = 0; i < N_columns; ++i)
-    {
-      data[i] = (double *) calloc(n_entries,sizeof(double));
+  int numEntries = 0;
+  FILE *fpi = fopen(filename,"r");
+  if (fpi) {
+    // Old code read the input file twice. Now we simply
+    // read all the lines from the input file into a string vector,
+    // then work with the data in-memory rather than do a second pass
+    // through the file.
+    // NB: LAMMPS coding guidelines prefer cstdio so we are intentionally
+    // foregoing  reading with getline
+    std::string message = fmt::format("INFO: About to read data file: {}", filename);
+    error->message(FLERR, message.c_str());
+
+    // Data file lines hold two floating point numbers.
+    // Line length we allocate should be long enough without being too long.
+    // 128 seems safe for a line we expect to be < 30 chars.
+    const int MAX_F_TABLE_LINE_LENGTH = 128;
+    char line[MAX_F_TABLE_LINE_LENGTH];
+    std::vector<std::string> inputLines;
+    while (fgets(line, MAX_F_TABLE_LINE_LENGTH, fpi)) {
+      inputLines.push_back(std::string(line));
     }
+    fclose(fpi);
+
+    numEntries = inputLines.size();
+    message = fmt::format("INFO: Read {} lines from file", numEntries);
+    error->message(FLERR, message.c_str());
 
-    // Don't need to re-open the file to make a second pass through it
-    // simply rewind to beginning
-    rewind(fpi);
+
+    // Initializing the size of vectors is not required.
+    // The structures will resize as you push values into them.
+    // But for largish data sets where you do know the value,
+    // it is more efficient to do one upfront allocation.
+    volumeVec.resize(numEntries);
+    pressureVec.resize(numEntries);
+    data.resize(NUM_INPUT_DATA_COLUMNS);
+    data[VOLUME] = volumeVec;
+    data[PRESSURE_CORRECTION] = pressureVec;
 
     double stdVolumeInterval = 0.0;
     double currVolumeInterval = 0.0;
     // When comparing doubles/floats, we need an Epsilon.
     // The literature indicates getting this value right in the
     // general case can be pretty complicated. I don't think it
     // needs to be complicated here, though. At least based on the
-    // sample data I've seen where the volume values are fairly
-    // large.
+    // sample data I've seen where the volume values are fairly large.
     const double volumeIntervalTolerance = 0.001;
-    n_entries = 0;
-    while( fgets(line,199,fpi)) {
-      ++n_entries;
-      test_sscanf = sscanf(line," %f , %f ",&f1, &f2);
+    int lineNum = 0;  // this value is only for  message
+    int numBadVolumeIntervals = 0; // count these for message
+    float f1, f2;
+    int test_sscanf;
+    for (int i = 0; i < inputLines.size(); ++i) {
+      lineNum++;  // count each line processed now so lineNum messages can be 1-based
+      test_sscanf = sscanf(inputLines.at(i).c_str()," %f , %f ",&f1, &f2);
```

Comment:
This should be using the ValueTokenizer class.

Reply:
Cool. I'll figure that out.

Reply:
@akohlmey Would you be open to my resolving all the other issues you've raised with this PR but deferring the conversion of the file-reading code to another pass? I'm more than happy to do that work but would like to get the fixed for the lost memory and invalid read merged and put to bed. My teaching semester approaches rapidly and I don't want the fixes for the lost memory to get stranded in my repo if it takes me a while to get to tthe file-reading changes.

Reply:
Yes, the refactoring of the file reading can wait.